main.py

```diff
@@ -866,23 +871,22 @@ def handle_link_down(self, event):  # pylint: disable=too-many-branches
                 switch_flows[dpid] = switch_flows[dpid][offset:]
             time.sleep(settings.BATCH_INTERVAL)
 
-        for evc in evcs_with_failover:
-            with evc.lock:
-                old_path = evc.current_path
-                evc.current_path = evc.failover_path
-                evc.failover_path = old_path
-                evc.sync()
-            emit_event(self.controller, "redeployed_link_down",
-                       content=map_evc_event_content(evc))
-            log.info(
-                f"{evc} redeployed with failover due to link down {link.id}"
-            )
-
         for evc in evcs_normal:
             emit_event(
                 self.controller,
                 "evc_affected_by_link_down",
-                content={"link_id": link.id} | map_evc_event_content(evc)
+                content={"link": link} | map_evc_event_content(evc)
+            )
+
+        for evc in evcs_with_failover:
+            evc.sync()
```

Comment:
@italovalcy, although with `evc.sync()` here now it's impacting less than what it would in prior commits, notice that it can still can take linearly `O(len(evcs_with_failover))`. Also, [`evc.setup_failover_path()`](https://github.com/kytos-ng/mef_eline/blob/issue_405/main.py#L896-L899) will make requests synchronously proportional to `O(len(check_failover))`, which in a scenario with 100+ EVCs can delay many seconds, for instance on multi topology with 100 EVCs it's taking 10+ secs:

![20240222_164715](https://github.com/kytos-ng/mef_eline/assets/1010796/3431e030-3c8c-4662-bc94-b66f1e2ad556)

That said, it might be worth completely deferring these side effects to be executed via another events (same idea how the other `emit_event` are structured).

Reply:
Italo, I believe after deferring more async writes I'll work. Ultimately though, what's making things harder is that actually what we optimally needed here was `evcs.bulk_write([UpdateOne(...), UpdateOne(...)])`  to update a set of EVCs. Now, regarding the two points I raised:

1 - Introducing a new method on `ELineController` something like `upsert_evcs(self, evcs: list[dict]) -> dict:` - a bulk version of `evc.sync()` - and letting MongoDB to efficiently execute these multiple updates. So, with this calling it directly from this handler would likely perform well (would need confirmation in pratice, but in theory it should be).

2 - Having an optional version for `evc.setup_failover_path()` in bulk would be a bit harder since this method is encapsulating many side effects, so just to isolate the path releated DB parts would be a large refactoring so it's not worth it. That said, maybe we could also think about another approach to only set up a new failover entirely asynchronously after some time? For example, when the network has been stable (no link down) after X configurable ms or so, wdyt, would that be acceptable? I think this can also bring more stability and avoid extra computations and the expense of not being optimal when handling a subsequent link failure. I mean comparing this with a traditional fast-reroute, usually a failover strategy can only guarantee optimal protection to a certain extent number of failures.

Reply:
Just some thoughts trying to think about options to bring more stability and minimize IO where applicable.